pkg/pool-weighted/contracts/smart/LiquidityBootstrappingPool.sol

```diff
@@ -59,73 +76,352 @@ contract LiquidityBootstrappingPool is BaseWeightedPool, ReentrancyGuard {
         _require(normalizedSum == FixedPoint.ONE, Errors.NORMALIZED_WEIGHT_INVARIANT);
 
-        // Initial value; can change later if weights change
-        _maxWeightTokenIndex = maxWeightTokenIndex;
+        // Write initial pool state - all zeros except for the start weights
+        _poolState = poolState;
+
+        // If false, the sale will start in a paused state (prevents front-running the unpause transaction)
+        swapEnabled = swapEnabledOnStart;
+
+        emit PublicSwapSet(swapEnabledOnStart);
+    }
+
+    // External functions
+
+    /**
+     * @dev Return start time, end time, and endWeights as an array
+     */
+    function getGradualUpdateParams()
+        external
+        view
+        returns (
+            uint256 startTime,
+            uint256 endTime,
+            uint256[] memory endWeights
+        )
+    {
+        // Load current pool state from storage
+        bytes32 poolState = _poolState;
+
+        startTime = poolState.decodeUint32(_START_TIME_OFFSET);
+        endTime = poolState.decodeUint32(_END_TIME_OFFSET);
+        uint256 totalTokens = _getTotalTokens();
+        endWeights = new uint256[](totalTokens);
+
+        // prettier-ignore
+        {
+            if (totalTokens > 0) { endWeights[0] = poolState.uncompress16(_END_WEIGHT_OFFSET); }
+            if (totalTokens > 1) { endWeights[1] = poolState.uncompress16(_END_WEIGHT_OFFSET + 16); }
+            if (totalTokens > 2) { endWeights[2] = poolState.uncompress16(_END_WEIGHT_OFFSET + 32); }
+            if (totalTokens > 3) { endWeights[3] = poolState.uncompress16(_END_WEIGHT_OFFSET + 48); }
+        }
+    }
+
+    /**
+     * @dev Can pause/unpause trading
+     */
+    function setPublicSwap(bool publicSwap) external authenticate whenNotPaused nonReentrant {
+        swapEnabled = publicSwap;
+
+        emit PublicSwapSet(publicSwap);
     }
 
+    /**
+     * @dev Schedule a gradual weight change, from the current weights to the given endWeights,
+     * over startTime to endTime
+     */
+    function updateWeightsGradually(
+        uint256 startTime,
+        uint256 endTime,
+        uint256[] memory endWeights
+    ) external authenticate whenNotPaused nonReentrant {
+        // solhint-disable-next-line not-rely-on-time
+        uint256 currentTime = block.timestamp;
+
+        _require(currentTime < endTime, Errors.GRADUAL_UPDATE_TIME_TRAVEL);
+
+        // Must specify normalized weights for all tokens
+        uint256 totalTokens = _getTotalTokens();
+        InputHelpers.ensureInputLengthMatch(totalTokens, endWeights.length);
+
+        // If the start time is in the past, "fast forward" to start now
+        // This avoids discontinuities in the weight curve. Otherwise, if you set the start/end times with
+        // only 10% of the period in the future, the weights would immediately jump 90%
+        uint256 effectiveStartTime = currentTime > startTime ? currentTime : startTime;
```

Comment:
This is from V1. I suppose it could be because of gas too, but the usual use case for this was when people wanted to reissue the same command they did before to just adjust the end weights, so they wouldn't have to recalculate anything, and so that the GUI would still show the original start time.